<a href="https://colab.research.google.com/github/jcjimenezb123/ProgramacionPython/blob/master/06_RegresionLineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Regresión Lineal con numpy

In [1]:
import numpy as np

In [2]:
x = np.array([0,1,2,3,4,5])
y = np.array([0,0.8,0.9,0.1,-0.8,-1.0])

In [3]:
pol = np.polyfit(x,y,1)
pol

array([-0.30285714,  0.75714286])

Ecuacion de la linea recta de la regresion lineal

$$
y_{cor}=-0.30285714x +  0.75714286
$$

In [4]:
y_cor = np.polyval(pol,x)
y_cor

array([ 0.75714286,  0.45428571,  0.15142857, -0.15142857, -0.45428571,
       -0.75714286])

In [6]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor,mode='lines+markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion Lineal',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

In [7]:
sse=sum((y-y_cor)**2)      # variabilidad explicada por la regresion
tse=(len(y))*np.var(y)   # variabilidad total de los datos
r2=1-sse/tse               # medida relativa entre las variabilidades
print(r2)

0.5177880184331798


##Regresion cuadratica

In [8]:
pol2 = np.polyfit(x,y,deg=2)
pol2

array([-0.16071429,  0.50071429,  0.22142857])

La ecuacion que representa la regresion cuadratica es

$$
y_{cor}=-0.16071429x^2+  0.50071429x+  0.22142857
$$

In [9]:
y_cor2 = np.polyval(pol2,x)
y_cor2

array([ 0.22142857,  0.56142857,  0.58      ,  0.27714286, -0.34714286,
       -1.29285714])

In [10]:
sse=sum((y-y_cor2)**2)
tse=(len(y)-1)*np.var(y)
r22=1-sse/tse
print(r22)

0.7946175115207375


In [15]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor2,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion cuadratica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

In [12]:
pol3 = np.polyfit(x,y,deg=3)
pol3

array([ 0.08703704, -0.81349206,  1.69312169, -0.03968254])

Usando una regresion cubica tenemos la ecuacion

$$
y_{cor}=0.08703704x^3 -0.81349206x^2+  1.69312169x  -0.03968254
$$

In [13]:
y_cor3=np.polyval(pol3,x)
y_cor3

array([-0.03968254,  0.92698413,  0.78888889,  0.06825397, -0.71269841,
       -1.03174603])

In [14]:
sse=sum((y-y_cor3)**2)
tse=(len(y)-1)*np.var(y)
r23=1-sse/tse
print(r23)

0.9846390168970814


In [17]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor3,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion cubica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

**Problema** En la fabricación de láminas de acero, el método para deformar acero a temperatura normal mantiene una relación inversa con la dureza del
mismo, ya que a medida que la deformación crece, se afecta la dureza del acero. Con el objetivo de conocer ésta relación se tomaron los siguientes datos

$$
\begin{array}{l|c|c|c|c|c|c|c|c|c|}
\text{Deformacion}&6&9&11&13&22&26&28&33&35\\
\hline
\text{Dureza}&68&67&65&53&44&40&37&34&32
\end{array}
$$

In [18]:
x = np.array([6,9,11,13,22,26,28,33,35])
y = np.array([68,67,65,53,44,40,37,34,32])

In [19]:
pol = np.polyfit(x,y,deg=1)
pol

array([-1.31956215, 75.71998588])

La ecuacion que relaciona la dureza con la deformacion es

$$
Dureza=-1.31956215Deformacion+ 75.71998588
$$

In [20]:
y_cor = np.polyval(pol,x)
y_cor

array([67.80261299, 63.84392655, 61.20480226, 58.56567797, 46.68961864,
       41.41137006, 38.77224576, 32.17443503, 29.53531073])

In [21]:
sse=sum((y-y_cor)**2)
tse=(len(y)-1)*np.var(y)
r2=1-sse/tse
print(r2)

0.949561766039208


In [23]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor,mode='lines+markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion Lineal',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

**Problema 2** Un proyectil es disparado hacia arriba desde el piso, se mide la altura del proyectil en el tiempo y se obtienen los siguientes datos.

$$
\begin{array}{l|c|c|c|c|c|c|}
\text{Tiempo(s)}& 0& 0.5& 1.0& 1.5& 2.0 &2.5\\
\hline
\text{Altura(ft)}& 0 &20.5& 31.36& 36.25& 30.41& 28.23
\end{array}
$$

In [24]:
x = np.array([0,0.5,1.0,1.5,2.0,2.5])
y = np.array([0,20.5,31.36,36.25,30.41,28.23])

In [25]:
pol2=np.polyfit(x,y,deg=2)
pol2

array([-12.87142857,  42.22257143,   1.17714286])

La ecuacion del tiro parabolico es

$$
Altura=-12.87142857t^2 + 42.22257143t  + 1.17714286
$$

In [26]:
y_cor2 = np.polyval(pol2,x)
y_cor2

array([ 1.17714286, 19.07057143, 30.52828571, 35.55028571, 34.13657143,
       26.28714286])

In [27]:
np.roots(pol2)

array([ 3.30797943, -0.02764647])

In [28]:
xs = np.linspace(x.min(),3.30)
ys = np.polyval(pol2,xs)

In [30]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor2,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))
traza3=go.Scatter(x=xs,y=ys,mode='lines',name='regresion',
                  line=dict(color='cyan',dash='dash'))

data=[traza1,traza2,traza3]

layout=go.Layout(title='Regresion cuadratica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

**Problema 3** Con los siguientes datos de Cp del Propano vs Temperatura, obtenga los coeficientes de la ecuación de Cp de la forma.

$$
Cp =a+bT+cT^2+dT^3
$$

Donde $a,b,c,d$ son los coeficientes a determinar y $T$ es la temperatura del Propano.

$$
\begin{array}{l|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|}
\text{Temp(C)}& 50 &100 &150 &200 &273.16 &298.15 &300& 400 &500 &600 &700 &800 &900 &1000 &1100\\
\hline
\text{Cp(kJ/kg)} &34.06 &41.3 &48.79 &56.07 &68.74 &73.6 &73.93 &94.01&112.59& 142.67 &154.77& 163.35 &174.6& 182.67& 128.7
\end{array}
$$

Aplique la regresion cubica para obtener los coeficientes $a,b,c,d.$

In [31]:
x = np.array([50 ,100 ,150 ,200 ,273.16 ,298.15 ,300, 400 ,500 ,600 ,700 ,800 ,900 ,1000 ,1100])
y = np.array([34.06,41.3,48.79 ,56.07 ,68.74,73.6,73.93,94.01,112.59,142.67,154.77,163.35,174.6,182.67,189.74])

In [32]:
pol3=np.polyfit(x,y,deg=3)
pol3

array([-1.57390652e-07,  2.00985608e-04,  1.14380718e-01,  2.70935513e+01])

$$
Cp=-1.57390652\times10^{-07}T^3+ 2.00985608\times10^{-04}T^2+1.14380718\times10^{-01}T+  27.0935513
$$

In [33]:
y_cor3 = np.polyval(pol3,x)
y_cor3

array([ 33.2953774 ,  40.38408853,  48.24164171,  56.74999397,
        70.12663817,  74.89104609,  75.24692374,  94.93053394,
       114.85648064, 134.08041994, 151.65800791, 166.64490065,
       178.09675424, 185.06922476, 186.6179683 ])

In [34]:
xs=np.linspace(x.min(),x.max())
ys=np.polyval(pol3,xs)

In [35]:
traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor3,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))
traza3=go.Scatter(x=xs,y=ys,mode='lines',name='regresion',
                  line=dict(color='cyan',dash='dash'))

data=[traza1,traza2,traza3]

layout=go.Layout(title='Regresion cubica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()